plugins/inputs/aos/aos.go

```diff
@@ -336,261 +402,351 @@ func (ssl *StreamAos) MsgReader(r io.Reader) {
+			for _, f := range newGenericPerfMon.GetFields() {
+				fName := f.GetName()
+				fValue := f.GetValue()
+
+				myValueOfValue := reflect.ValueOf(fValue).Elem()
+				myType := myValueOfValue.Type().String()
+
+				switch myType {
+				case "aos_streaming.Field_FloatValue":
+					fields[fName] = f.GetFloatValue()
+				case "aos_streaming.Field_Int64Value":
+					fields[fName] = f.GetInt64Value()
+				case "aos_streaming.Field_StringValue":
+					log.Printf("W! Perfmon Generic - Field %v can't be of type String, must be Float of Int64", fName)
+				}
 			}
 
-			if newGenericPerfMon != nil {
-
-				serie := "perfmon_generic_undefined"
-				fields := make(map[string]interface{})
-				tags := ssl.GetTags(originName)
-
-				for _, t := range newGenericPerfMon.GetTags() {
-					tName := t.GetName()
-					tValue := t.GetValue()
-
-					myValueOfName := reflect.ValueOf(tValue).Elem()
-					myType := myValueOfName.Type().String()
-
-					// Intercept the special tag "data_type"
-					if tName == "data_type" {
-						serie = t.GetStringValue()
-						continue
-					}
-
-					switch myType {
-					case "aos_streaming.Tag_StringValue":
-						tags[tName] = t.GetStringValue()
-					case "aos_streaming.Tag_FloatValue":
-						log.Printf("W! Perfmon Generic - Tag can only be of type String, %v is type Float", tName)
-					case "aos_streaming.Tag_Int64Value":
-						log.Printf("W! Perfmon Generic - Tag can only be of type String, %v is type Int64", tName)
-					}
-				}
-				for _, f := range newGenericPerfMon.GetFields() {
-					fName := f.GetName()
-					fValue := f.GetValue()
-
-					myValueOfValue := reflect.ValueOf(fValue).Elem()
-					myType := myValueOfValue.Type().String()
-
-					switch myType {
-					case "aos_streaming.Field_FloatValue":
-						fields[fName] = f.GetFloatValue()
-					case "aos_streaming.Field_Int64Value":
-						fields[fName] = f.GetInt64Value()
-					case "aos_streaming.Field_StringValue":
-						log.Printf("W! Perfmon Generic - Field %v can't be of type String, must be Float of Int64", fName)
-					}
-				}
+			ssl.Aos.Accumulator.AddFields(serie, fields, tags)
+		}
+	}
 
-				ssl.Aos.Accumulator.AddFields(serie, fields, tags)
+	if newEvent != nil {
+
+		// ----------------------------------------------------------------
+		// Collect all type of events
+		// ----------------------------------------------------------------
+		if ssl.IsSequencedStream {
+			ssl.eventSeqInit.Do(func() {
+				ssl.EventSequence = sequencedMsg.GetSeqNum()
+				ssl.EventSeqGap = 0
+				ssl.reportMessageLoss("event", 0, 0)
+			})
+			seqGap := sequencedMsg.GetSeqNum() - ssl.EventSequence
+			if seqGap != ssl.EventSeqGap {
+				log.Printf("W! Event sequence number expected %d got %d", ssl.EventSequence, sequencedMsg.GetSeqNum())
+				ssl.reportMessageLoss("event", ssl.EventSequence, sequencedMsg.GetSeqNum())
+				ssl.EventSequence = sequencedMsg.GetSeqNum() + 1
+				ssl.EventSeqGap = seqGap
+			} else {
+				ssl.EventSequence++
```

Comment:
This logic can be made as a common function and can be reused across all the streams.

Reply:
I'll take a look, but my concern is that this function will need to be thread-safe.

Reply:
I'm not sure about this. The sequence numbers for each type are tracked in separate variables. I think most of this code will end up the same but in a function with a case statement that also needs to be thread-safe.